File: Algorithm-Python/word-search.py
class Solution(object):
    def existRecursive(self, board, word, row, col,nrow,ncol, ind,nword):

        if board[row][col] == word[ind]:
            temp = board[row][col]
            board[row][col] = '*'
            if ind == nword - 1:
                return True
            elif (row + 1 < nrow and board[row + 1][col] != '*' and self.existRecursive(board, word, row + 1, col,nrow,ncol,
                                                                               ind + 1,nword)) or (
                                    row - 1 >= 0 and board[row - 1][col] != '*' and self.existRecursive(board, word, row - 1,
                                                                                               col,nrow,ncol,
                                                                                               ind + 1,nword)) or (
                                    col + 1 < ncol and board[row][col + 1] != '*' and self.existRecursive(board, word, row,
                                                                                                 col + 1,nrow,ncol,
                                                                                                 ind + 1,nword)) or (
                                    col - 1 >= 0 and board[row][col - 1] != '*' and self.existRecursive(board, word, row,
                                                                                               col - 1,nrow,ncol,
                                                                                               ind + 1,nword)):
                return True
            board[row][col] = temp
        else:
            return False

    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        nrow = len(board)
        ncol = len(board[0])
        nword = len(word)

        for row in range(0, nrow):
            for col in range(0, ncol):
                if board[row][col] == word[0] and self.existRecursive(board, word, row, col,nrow,ncol, 0,nword):
                    return True
        return False


s = Solution()

File: Algorithm-CPP/Algorithm-C/word-search.c
//
//  word-search.c
//
//
//  Created by Sanqiang Zhao on 11/25/15.
//
//

#include "word-search.h"
bool existRecursive(char** board, int boardRowSize, int boardColSize, char* word, int idx, int nword,int row, int col){
    if (board[row][col] == word[idx]) {
        char temp =board[row][col];
        board[row][col] = '*';
        if (idx == nword-1) {
            return true;
        }else if((row +1 < boardRowSize && board[row+1][col]!='*' && existRecursive(board,boardRowSize,boardColSize,word,idx+1,nword,row+1,col))  ||(row -1 >=0 && board[row-1][col]!='*' && existRecursive(board,boardRowSize,boardColSize,word,idx+1,nword,row-1,col)) ||(col +1 < boardColSize && board[row][col+1]!='*' && existRecursive(board,boardRowSize,boardColSize,word,idx+1,nword,row,col+1))  ||(col -1 >=0 && board[row][col-1]!='*' && existRecursive(board,boardRowSize,boardColSize,word,idx+1,nword,row,col-1))){
            return true;
        }
        board[row][col] =temp;
    }
    return false;
}


bool exist(char** board, int boardRowSize, int boardColSize, char* word) {
    unsigned long nword = strlen(word);
    
    for (int row = 0; row < boardRowSize; ++row) {
        for (int col = 0; col < boardColSize ; ++col) {
            if(word[0] == board[row][col] && existRecursive(board,boardRowSize,boardColSize,word,0,nword,row,col)){
                return true;
            }
        }
    }
    return false;
}